tests/MongoDB.Driver.Tests/Specifications/crud/CrudTestRunner.cs

```diff
@@ -19,55 +19,228 @@
+
         [SkippableTheory]
         [ClassData(typeof(TestCaseFactory))]
         public void RunTestDefinition(BsonDocument definition, BsonDocument test, bool async)
         {
-            BsonValue minServerVersion;
-            if (definition.TryGetValue("minServerVersion", out minServerVersion))
+            SkipTestIfNeeded(definition, test);
+
+            var collectionName = GetCollectionName(definition);
+            DropCollection(DriverTestConfiguration.DatabaseNamespace.DatabaseName, collectionName);
+            PrepareData(DriverTestConfiguration.DatabaseNamespace.DatabaseName, collectionName, definition);
+
+            using (var client = CreateDisposableClient(_capturedEvents))
             {
-                RequireServer.Check().VersionGreaterThanOrEqualTo(minServerVersion.AsString);
+                var database = client.GetDatabase(DriverTestConfiguration.DatabaseNamespace.DatabaseName);
+                var operations = test.GetValue("operations", new BsonArray()).AsBsonArray;
+                if (operations.Count <= 0) // Assuming that we cannot have `operations` and `operation` at the same time
+                {
+                    operations.Add(test["operation"]);
+                }
+
+                foreach (var operation in operations.Select(c => c.AsBsonDocument))
+                {
+                    var collection = GetCollection(database, collectionName, operation);
+                    ExecuteOperation(client, database, collection, operation, (BsonDocument)test.GetValue("outcome", null), async);
+                }
+
+                AssertEventsIfNeeded(_capturedEvents, test);
             }
+        }
 
-            BsonValue maxServerVersion;
-            if (definition.TryGetValue("maxServerVersion", out maxServerVersion))
+        private void AssertEvent(object actualEvent, BsonDocument expectedEvent)
+        {
+            if (expectedEvent.ElementCount != 1)
             {
-                RequireServer.Check().VersionLessThanOrEqualTo(maxServerVersion.AsString);
+                throw new FormatException("Expected event must be a document with a single element with a name the specifies the type of the event.");
             }
 
-            var database = DriverTestConfiguration.Client
-                .GetDatabase(DriverTestConfiguration.DatabaseNamespace.DatabaseName);
-            var collection = database
-                .GetCollection<BsonDocument>(DriverTestConfiguration.CollectionNamespace.CollectionName);
+            var eventType = expectedEvent.GetElement(0).Name;
+            var eventAsserter = EventAsserterFactory.CreateAsserter(eventType);
+            eventAsserter.AssertAspects(actualEvent, expectedEvent[0].AsBsonDocument);
+        }
+
+        private void AssertEventsIfNeeded(EventCapturer actualEvents, BsonDocument test)
+        {
+            if (test.Contains("expectations"))
+            {
```

Comment:
It could also be refactored to not use enumerators at all. We can use a foreach loop to compare the actual events to the expected events and then test whether there are more actual events than expected events or vice versa:

    var actualEvents = eventCapturer.Events;
    var expectedEvents = test["expectations"].AsBsonArray.Cast<BsonDocument>().ToList();

    var minCount = Math.Min(actualEvents.Count, expectedEvents.Count);
    for (var i = 0; i < minCount; i++)
    {
        AssertEvent(actualEvents[i], expectedEvents[i]);
    }

    if (actualEvents.Count < expectedEvents.Count)
    {
        throw new Exception($"Missing event: {expectedEvents[actualEvents.Count]}.");
    }

    if (actualEvents.Count > expectedEvents.Count)
    {
        throw new Exception($"Unexpected event of type: {actualEvents[expectedEvents.Count].GetType().Name}.");
    }

Reply:
Done